### ISSUE TITLE
clippy::eval_order_dependence leaks from macro and is too sensitive

### ISSUE THREAD

**Issue by Tamschi** [rust-lang/rust-clippy#8484]
### Summary

I wrote a frontend framework that flows the dependency injection context of logical containers to their content children. It does this through expressions of the following shape:

```rust
let c = …;

let c_0;
let c_1;
Parent {
    container: {
        let constructed = Container::new(c.as_ref());
        c_0 = constructed.1;
        constructed.0
    },
    content: {
        let constructed = Content::new(c_0.as_ref());
        c_1 = constructed.1;
        constructed.0
    },
}
```

From the expansion I can see that the lint is attached to the `c_0` in `c_0.as_ref()`, which has `Span::mixed_site()` resolution everywhere it appears and as such should be suppressed in the input.
(I use `Span`-bending to map the location to an input token instead, to make it easier to debug these issues. This does not seem to affect whether other lints are suppressed.)

- - -

Additionally, **I believe this lint should not flag cases where there is only a single assignment to a previously unassigned binding**.
(I suppose that's my main and most actionable point here, hence the highlight.)

The alternative in this case would be to construct and bind everything via let bindings in advance which, at least in my eyes, would make more deeply nested expressions harder to follow. (There are various expressions in my template grammar that require a separate nested storage context, i.e. a nested generated and usually generated struct type, for therein contained child components and field bindings.)

### Lint Name

eval_order_dependence

### Reproducer

Since the actual example for the macro leak is somewhat extensive, I created a branch in my repository for the issue report here: https://github.com/Tamschi/Asteracea/tree/reports/clippy--eval_order_dependence

The test `dependency_injection_by_container_` is the expanded form that shows the location of the warning in more detail ([in this revision and line](https://github.com/Tamschi/Asteracea/blob/3e18b4e1216a5c34586afb21af6235712a2c28f9/tests/dependency_injection_by_container_.rs#L948)).

`dependency_injection_by_container` is one of the tests where the warning leaks out from the macro.

### Version

```text
rustc 1.61.0-nightly (532d3cda9 2022-02-23)
binary: rustc
commit-hash: 532d3cda90b8a729cd982548649d32803d265052
commit-date: 2022-02-23
host: x86_64-pc-windows-msvc
release: 1.61.0-nightly
LLVM version: 14.0.0
```

**Comment by Tamschi** [comment]
I seems I can't easily suppress this for just that expression either, since writing

```rust
{
	#![allow(clippy::eval_order_dependence)] // Warning is too strict.
	#resource_node.as_ref()
},
```

(or variations with outer attributes and/or without that block) has no effect.

**Comment by Jarcho** [comment]
The lint has been renamed to `mixed_read_write_expression` as of #8621.
